File: constitutional/runtime/constitutional_state.py
"""Article XV — Constitutional State Runtime (receipt-v2 bridge over constitutional_state core)."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Literal, TypeAlias

from pydantic import BaseModel, Field

from constitutional.core.graph import DOMAIN_STATE_MAPS, LEGAL_TRANSITIONS, map_domain_state, validate_transition
from constitutional.core.models import StateObject, Transition
from constitutional.runtime.receipts_v2 import (
    TransitionReceiptV2,
    is_receipt_v2_complete,
    stable_json_hash,
)

StateTransition = Transition

ConstitutionalStateName: TypeAlias = Literal[
    "Proposed",
    "Evaluated",
    "Approved",
    "Executed",
    "Observed",
    "Challenged",
    "Arbitrated",
    "Remediated",
    "Closed",
]

StateObjectType: TypeAlias = Literal[
    "ClaimState",
    "AuthorityState",
    "InstitutionState",
    "DecisionState",
    "ContinuityState",
    "SovereigntyState",
    "RealityState",
    "DomainState",
]


class ReplayResult(BaseModel):
    reconstructed_state: str
    canonical_state: str
    diverged: bool
    history_length: int
    reconstructed_version: int
    canonical_version: int
    state_id: str
    replay_hash: str


def transition_from_receipt(
    receipt: TransitionReceiptV2,
    *,
    state_object_id: str,
    accountable_party: str,
) -> Transition:
    if not is_receipt_v2_complete(receipt):
        raise ValueError(f"incomplete transition receipt: {receipt.receipt_id}")
    ts = receipt.timestamp
    try:
        timestamp = datetime.fromisoformat(ts.replace("Z", "+00:00"))
    except ValueError:
        timestamp = datetime.now(timezone.utc)
    return Transition(
        transition_id=receipt.receipt_id,
        state_object_id=state_object_id,
        from_state=receipt.transition.from_state,
        to_state=receipt.transition.to_state,
        receipt_id=receipt.receipt_id,
        runtime=receipt.runtime,
        legal_basis=receipt.transition.legal_basis,
        accountable_party=accountable_party,
        timestamp=timestamp,
    )


def reconstruct_state(
    receipts: list[TransitionReceiptV2],
    state_obj: StateObject,
    *,
    accountable_party: str = "operator",
) -> StateObject:
    working = state_obj.model_copy(deep=True)
    for receipt in receipts:
        if receipt.transition.state_id and receipt.transition.state_id != working.state_id:
            raise ValueError(
                f"receipt {receipt.receipt_id} targets state {receipt.transition.state_id}, "
                f"expected {working.state_id}"
            )
        working.apply_transition(
            transition_from_receipt(
                receipt,
                state_object_id=working.state_id,
                accountable_party=accountable_party,
            )
        )
    return working


def reconstruct_state_at(
    receipts: list[TransitionReceiptV2],
    state_obj: StateObject,
    *,
    at_index: int,
    accountable_party: str = "operator",
) -> StateObject:
    if at_index < 0:
        raise ValueError("at_index must be >= 0")
    return reconstruct_state(receipts[: at_index + 1], state_obj, accountable_party=accountable_party)


def replay_state(
    receipts: list[TransitionReceiptV2],
    canonical_state: StateObject,
    *,
    accountable_party: str = "operator",
) -> ReplayResult:
    seed = StateObject(
        state_id=canonical_state.state_id,
        state_type=canonical_state.state_type,
        current_state="Proposed",
        invariants=list(canonical_state.invariants),
        evidence_requirements=list(canonical_state.evidence_requirements),
        authority_model=list(canonical_state.authority_model),
        reproducibility_requirements=list(canonical_state.reproducibility_requirements),
        impact_boundaries=list(canonical_state.impact_boundaries),
        accountability_chain=list(canonical_state.accountability_chain),
    )
    reconstructed = reconstruct_state(receipts, seed, accountable_party=accountable_party)
    diverged = (
        reconstructed.current_state != canonical_state.current_state
        or reconstructed.version != canonical_state.version
    )
    material = {
        "reconstructed": stable_json_hash(reconstructed.model_dump()),
        "canonical": stable_json_hash(canonical_state.model_dump()),
    }
    return ReplayResult(
        reconstructed_state=reconstructed.current_state,
        canonical_state=canonical_state.current_state,
        diverged=diverged,
        history_length=len(reconstructed.history),
        reconstructed_version=reconstructed.version,
        canonical_version=canonical_state.version,
        state_id=canonical_state.state_id,
        replay_hash=stable_json_hash(material),
    )


def parse_transition_receipts(
    payloads: list[dict | TransitionReceiptV2],
) -> list[TransitionReceiptV2]:
    out: list[TransitionReceiptV2] = []
    for item in payloads:
        if isinstance(item, TransitionReceiptV2):
            out.append(item)
        else:
            out.append(TransitionReceiptV2.model_validate(item))
    return out

File: aaes-os/packages/governed-memory/src/constitutional_state.ts
/** Article XV — Constitutional State Runtime (shared substrate of governed state). */

import type { TransitionReceiptV2 } from './receipts_v2.js';
import { isReceiptV2Complete } from './receipts_v2.js';

export type ConstitutionalStateName =
  | 'Proposed'
  | 'Evaluated'
  | 'Approved'
  | 'Executed'
  | 'Observed'
  | 'Challenged'
  | 'Arbitrated'
  | 'Remediated'
  | 'Closed';

export type StateObjectType =
  | 'ClaimState'
  | 'AuthorityState'
  | 'InstitutionState'
  | 'DecisionState'
  | 'ContinuityState'
  | 'SovereigntyState'
  | 'RealityState'
  | 'DomainState';

export const LEGAL_TRANSITIONS: Record<string, ConstitutionalStateName[]> = {
  Proposed: ['Evaluated'],
  Evaluated: ['Approved'],
  Approved: ['Executed'],
  Executed: ['Observed'],
  Observed: ['Challenged', 'Closed'],
  Challenged: ['Arbitrated'],
  Arbitrated: ['Remediated'],
  Remediated: ['Closed'],
};

export const DOMAIN_STATE_MAPS: Record<string, Record<string, ConstitutionalStateName>> = {
  Truth: {
    Supported: 'Evaluated',
    Verified: 'Approved',
    Diverged: 'Challenged',
  },
  Sovereignty: {
    Requested: 'Proposed',
    Delegated: 'Approved',
    Active: 'Executed',
    Suspended: 'Challenged',
    Revoked: 'Remediated',
  },
  Institutional: {
    Draft: 'Proposed',
    Audited: 'Challenged',
    Amended: 'Remediated',
  },
  Continuity: {
    EventRecorded: 'Executed',
  },
  Reproduction: {
    Reproduced: 'Executed',
    Diverged: 'Challenged',
  },
};

export interface StateTransition {
  transition_id: string;
  from_state: string;
  to_state: string;
  receipt_id: string;
  timestamp: string;
  legal_basis?: string;
  receipt_ids_used?: string[];
}

export interface StateObject {
  state_id: string;
  state_type: string;
  version: number;
  current_state: string;
  invariants: string[];
  evidence_requirements: string[];
  authority_model: string[];
  reproducibility_requirements: string[];
  impact_boundaries: string[];
  accountability_chain: string[];
  history: StateTransition[];
}

export interface ReplayResult {
  reconstructed_state: string;
  canonical_state: string;
  diverged: boolean;
  history_length: number;
  reconstructed_version: number;
  canonical_version: number;
  state_id: string;
  replay_hash: string;
}

export function validateTransition(fromState: string, toState: string): void {
  const allowed = LEGAL_TRANSITIONS[fromState] ?? [];
  if (!allowed.includes(toState as ConstitutionalStateName)) {
    throw new Error(`Illegal transition: ${fromState} → ${toState}`);
  }
}

export function mapDomainState(domain: string, domainState: string): string {
  return DOMAIN_STATE_MAPS[domain]?.[domainState] ?? domainState;
}

export function transitionFromReceipt(receipt: TransitionReceiptV2): StateTransition {
  if (!isReceiptV2Complete(receipt)) {
    throw new Error(`incomplete transition receipt: ${receipt.receipt_id}`);
  }
  return {
    transition_id: receipt.receipt_id,
    from_state: receipt.transition.from_state,
    to_state: receipt.transition.to_state,
    receipt_id: receipt.receipt_id,
    timestamp: receipt.timestamp,
    legal_basis: receipt.transition.legal_basis,
    receipt_ids_used: [...receipt.transition.receipt_ids_used],
  };
}

export function applyTransition(state: StateObject, transition: StateTransition): StateObject {
  if (transition.from_state !== state.current_state) {
    throw new Error(
      `Illegal transition for ${state.state_id}: ${state.current_state} → ${transition.to_state}`,
    );
  }
  validateTransition(transition.from_state, transition.to_state);
  return {
    ...state,
    current_state: transition.to_state,
    version: state.version + 1,
    history: [...state.history, transition],
  };
}

export function reconstructState(
  receipts: TransitionReceiptV2[],
  stateObj: StateObject,
): StateObject {
  let working = { ...stateObj, history: [...stateObj.history] };
  for (const receipt of receipts) {
    const transition = transitionFromReceipt(receipt);
    if (receipt.transition.state_id && receipt.transition.state_id !== working.state_id) {
      throw new Error(
        `receipt ${receipt.receipt_id} targets state ${receipt.transition.state_id}, expected ${working.state_id}`,
      );
    }
    working = applyTransition(working, transition);
  }
  return working;
}

export function replayState(
  receipts: TransitionReceiptV2[],
  canonicalState: StateObject,
): ReplayResult {
  const seed: StateObject = {
    state_id: canonicalState.state_id,
    state_type: canonicalState.state_type,
    version: 0,
    current_state: 'Proposed',
    invariants: [...canonicalState.invariants],
    evidence_requirements: [...canonicalState.evidence_requirements],
    authority_model: [...canonicalState.authority_model],
    reproducibility_requirements: [...canonicalState.reproducibility_requirements],
    impact_boundaries: [...canonicalState.impact_boundaries],
    accountability_chain: [...canonicalState.accountability_chain],
    history: [],
  };
  const reconstructed = reconstructState(receipts, seed);
  const diverged =
    reconstructed.current_state !== canonicalState.current_state ||
    reconstructed.version !== canonicalState.version;
  return {
    reconstructed_state: reconstructed.current_state,
    canonical_state: canonicalState.current_state,
    diverged,
    history_length: reconstructed.history.length,
    reconstructed_version: reconstructed.version,
    canonical_version: canonicalState.version,
    state_id: canonicalState.state_id,
    replay_hash: `${reconstructed.current_state}:${reconstructed.version}:${canonicalState.current_state}:${canonicalState.version}`,
  };
}
